servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,547 @@
+            private volatile int pendingOnNext;
+            /**
+             * visibility provided by the {@link Subscriber} thread in {@link #onNext(Object)}, and then by
+             * {@link #pendingDemand} when written to from {@link #replenishDemandAndClearSignalsQueued()} (potentially
+             * called by a different thread). Demand is exhausted before {@link #replenishDemandAndClearSignalsQueued()}
+             * is called, and that method triggers {@link #request(long)} and
+             * <a href="https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.3/README.md#1.1">there’s a
+             * happens-before relationship between requesting elements and receiving elements</a>.
+             */
+            private boolean signalsQueued;
+
+            FlatMapPublisherSubscriber(FlatMapSubscriber<T, R> parent) {
+                this.parent = parent;
+                subscription = new DelayedSubscription();
+            }
+
+            void cancelFromUpstream() {
+                subscription.cancel();
+            }
+
+            void replenishDemandAndClearSignalsQueued() {
+                // clearSignalsQueued must be before pendingOnNext for visibility with onNext.
+                clearSignalsQueued();
+                final int prevPendingOnNext = pendingOnNextUpdater.getAndSet(this, 0);
+                if (prevPendingOnNext > 0) {
+                    subscription.request(prevPendingOnNext);
+                }
+            }
+
+            void clearSignalsQueued() {
+                signalsQueued = false;
+            }
+
+            void markSignalsQueued() {
+                signalsQueued = true;
+            }
+
+            boolean hasSignalsQueued() {
+                return signalsQueued;
+            }
+
+            @Override
+            public void onSubscribe(final Subscription s) {
+                subscription.delayedSubscription(ConcurrentSubscription.wrap(s));
+                subscription.request(parent.source.maxMappedDemand);
+            }
+
+            @Override
+            public void onNext(@Nullable final R r) {
+                // pendingOnNext must be updated before tryEmitItem for visibility of signalsQueued. signalsQueued
+                // is updated in replenishDemandAndClearSignalsQueued (potentially from another thread) before
+                // pendingOnNext is reset and we want signalsQueued to be visible to this thread.
+                //
+                // Heuristic which triggers replenishDemandAndClearSignalsQueued when demand is consumed. It must be
+                // when all demand is consumed in order for clearSignalsQueued to work properly.
+                final boolean doReplenish = pendingOnNextUpdater.incrementAndGet(this) == parent.source.maxMappedDemand;
+                parent.tryEmitItem(wrapNull(r), this);
+                if (doReplenish) {
+                    parent.tryEmitItem(this, this);
```

Comment:
This is done to invoke `replenishDemandAndClearSignalsQueued` after the exhausted demand has been processed to avoid requesting more than max concurrency (the state for demand is maintained in each `Subscriber`). Can you clarify the source of confusion and do you have a suggestion to improve?